adafruit_connection_manager.py

```diff
@@ -0,0 +1,402 @@
+            )
+
+        addr_info = self._socket_pool.getaddrinfo(
+            host, port, 0, self._socket_pool.SOCK_STREAM
+        )[0]
+
+        retry_count = 0
+        socket = None
+        last_exc = None
+        last_exc_new_type = None
+        while retry_count < max_retries and socket is None:
+            if retry_count > 0:
+                logger.debug(f"  retry #{retry_count}")
+                if any(self._socket_free.items()):
+                    self._free_sockets()
+                else:
+                    raise RuntimeError("Sending request failed") from last_exc
+            retry_count += 1
+
+            try:
+                socket = self._socket_pool.socket(addr_info[0], addr_info[1])
+            except OSError as exc:
+                logger.debug(f"  OSError getting socket: {exc}")
+                last_exc_new_type = SocketGetOSError
+                last_exc = exc
+                continue
+            except RuntimeError as exc:
+                logger.debug(f"  RuntimeError getting socket: {exc}")
+                last_exc_new_type = SocketGetRuntimeError
+                last_exc = exc
+                continue
+
+            connect_host = addr_info[-1][0]
+            if is_ssl:
+                socket = ssl_context.wrap_socket(socket, server_hostname=host)
+                connect_host = host
+            socket.settimeout(timeout)  # socket read timeout
+
+            try:
+                socket.connect((connect_host, port))
+            except MemoryError as exc:
+                logger.debug(f"  MemoryError connecting socket: {exc}")
+                last_exc_new_type = SocketConnectMemoryError
+                last_exc = exc
+                socket.close()
+                socket = None
+            except OSError as exc:
+                logger.debug(f"  OSError connecting socket: {exc}")
+                last_exc_new_type = SocketConnectOSError
+                last_exc = exc
+                socket.close()
+                socket = None
+
+        if socket is None:
+            logger.debug("  Repeated socket failures")
+            if exception_passthrough:
+                raise last_exc_new_type("Repeated socket failures") from last_exc
+            raise RuntimeError("Repeated socket failures") from last_exc
+
+        self._open_sockets[key] = socket
+        self._socket_free[socket] = False
+        return socket
+
+
+# connection helpers
+
+
+_global_connection_manager = None  # pylint: disable=invalid-name
+
+
+def get_connection_manager(socket_pool: SocketpoolModuleType) -> None:
+    """Get the ConnectionManager singleton"""
+    global _global_connection_manager  # pylint: disable=global-statement
+    if _global_connection_manager is None:
+        _global_connection_manager = ConnectionManager(socket_pool)
+    return _global_connection_manager
+
+
+def get_connection_members(radio):
+    """Helper to get needed connection mamers for common boards"""
```

Comment:
I'm not really a fan of this because `members` is vague. Could this be the ConnectionManager constructor instead and then have requests.Session et al recognize when they are given a connection manager?

Another clearer way would be to have two functions, `get_pool()` and `get_ssl_context()`.

Reply:
Can make this `get_connection_pool_and_ssl_context`. I had this idea of just passing the radio to things like requests and having it get the `pool` and `ssl_context`. The thought was around simplifying imports.
I can also take this out of here for now, and add this as a second add.

Reply:
This also allows you to:
```
requests = adafruit_requests.Session(*connection_members)
```
or even:
```
requests = adafruit_requests.Session(*get_connection_members(radio))
```

Reply:
I'm not a huge fan of what it is hiding. I'd prefer it in a follow up PR.

Reply:
I did break these into `get_radio_socketpool` and `get_radio_ssl_contexts`. Still happy to remove if we don't want theses yet